File: 백준/단지번호붙이기.py
from collections import deque

input=__import__('sys').stdin.readline
MIS=lambda:map(int,input().rstrip())
n=int(input().rstrip());board=[]
for _ in range(n):
    board.append(list(MIS()))

dx=[-1,0,1,0];dy=[0,-1,0,1]
visited=[[False for _ in range(n)] for _ in range(n)]
q=deque();danji=0
result=[]
for i in range(len(board)):
    for j in range(len(board[i])):
        if board[i][j]==1 and not visited[i][j]:
            danji+=1
            cnt=1
            q.append((i,j));visited[i][j]=True
            while q:
                x,y=q.popleft()
                for idx in range(4):
                    nx,ny=x+dx[idx],y+dy[idx]
                    if 0<=nx<n and 0<=ny<n and board[nx][ny]==1 and not visited[nx][ny]:
                        cnt+=1
                        visited[nx][ny]=True
                        q.append((nx,ny))
            result.append((danji,cnt)) 

print(len(result))
result=sorted(result,key=lambda x: x[1])
for danji,cnt in result:
    print(cnt)

File: 백준/java/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static int n;
	static int[][] board;
	static int[] dx = {-1,0,1,0};
	static int[] dy = {0,-1,0,1};
	static boolean[][] visited;
	static ArrayList<Integer> result = new ArrayList<>();
	static class Pos{
		int x;
		int y;
		Pos(int x, int y){
			this.x = x;
			this.y = y;
		}
	}
	
	public static int bfs() {
		Queue<Pos> q = new LinkedList<>();
		visited = new boolean[n][n];
		int group=0,cnt;
		for(int i=0; i<n; i++) {
			for(int j=0; j<n; j++) {
				if(board[i][j]==1 && !visited[i][j]) {
					q.offer(new Pos(i,j));
					visited[i][j]=true;
					group+=1;
					cnt=1;
					int nx,ny;
					while(!q.isEmpty()) {
						Pos now = q.poll();
						for(int idx=0; idx<4; idx++) {
							nx = now.x+dx[idx];
							ny = now.y+dy[idx];
							if(0<=nx && nx<n && 0<=ny && ny<n && !visited[nx][ny] && board[nx][ny]==1) {
								q.offer(new Pos(nx,ny));
								visited[nx][ny]=true;
								cnt+=1;
							}
						}
					}
					result.add(cnt);
				}
			}
		}
		
		return group;
	}
	
	public static void main(String[] args) throws IOException{
		// TODO Auto-generated method stub
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		n=Integer.parseInt(br.readLine());
		board = new int[n][n];
		for(int i=0; i<n; i++) {
			String data = br.readLine();
			for(int j=0; j<n; j++) {
				board[i][j] = data.charAt(j)-'0';
			}
		}
		
		System.out.println(bfs());
		Collections.sort(result);
		for(int num : result) {
			System.out.println(num);
		}
		
	}

}
